### ISSUE TITLE
NC22 Support

### ISSUE THREAD

**Issue by sunjam** [nursoda/twofactor_email#187]
Version 22 was just released.  Thanks for considering.

**Comment by gorzala** [comment]
I would also need NC22 Support. After upgrading my server to 22 I could not login to with my admin account.
(I could not disable the email provider with occ twofactorauth:disable marc email)

**Comment by TheThief** [comment]
It's trivial to modify the appinfo/info.xml to set maxversion to 22 - in fact it already is in latest code, it's just not been officially released.

**Comment by TheThief** [comment]
@rullzer please release the latest version with NC 22 support

**Comment by andyxheli** [comment]
yes please

**Comment by rullzer** [comment]
I no longer maintain this. If anybody wants to take over feel free to reach out.

**Comment by nursoda** [comment]
I PM'ed you (to your signoff-address).

**Comment by nursoda** [comment]
As you see from the redirection rullzer transferred this repo to me. I have updated all dev dependencies and merged the "update to 22" PR by rullzer along with another typor PR. I signed the updated packages and they look perfect to me. I hesitate to upload them to the app store since it says there that by uploading my CRT all prior apps will be deleted. I wait for confirmation from Nextcloud team folks. @users: Please be patient. @nextcloud-gmbh: Please advise.

**Comment by ChristophWurst** [comment]
If I'm not wrong then due to the update of the certificate revocation list shipped with the next Nextcloud releases the old releases will already become unusable. Just re-register the app, sign the releases with your key and re-release them.

**Comment by nursoda** [comment]
Thanks. All versions? What I did for the old versions was

1. download the current releases
2. extract them to my NC22 test installation (although they would not work there)
3. sign them using occ (which gave a warning on DB updated to be necessary which I ignored)
4. verify that the new signature file was written and re-pack them

I'll proceed now :) If all goes wrong, I'd need some interactive support :) 🤞🏻

**Comment by nursoda** [comment]
Didn't work, I cannot update the certificate in the app store since I get this error:

```
Only the app owner is allowed to update the certificate
```

@ChristophWurst: Will you take care and escalate this or who is the right person for me to ask?

**Comment by ChristophWurst** [comment]
https://github.com/nextcloud/appstore/search?q=%22Only+the+app+owner+is+allowed+to+update+the+certificate%22 -> https://github.com/nextcloud/appstore/blob/77e2d33f2318c493b6c38a10d1f81dfca8556339/nextcloudappstore/api/v1/views.py#L129

You are not registered as the owner of the app yet. I'll ask one of the admins to make you the owner of the app on https://apps.nextcloud.com/apps/twofactor_email.

**Comment by ChristophWurst** [comment]
You should be the owner now. Please retry.

**Comment by nursoda** [comment]
Yes, thanks for your help, it worked fine now. I have two more questions:

* Some of my clouds (those hosted by providers) did receive the update right away, but my self-hosted clouds don't see it yet. What could be the cause for that? Are cron jobs (on my systems or on the store) involved so that it's expected to take a while?
* I'd like to tinker the description of the app, like adding an issue URL, app icon/image, a screenshot, etc. Those may be set when ["generating" an app](https://apps.nextcloud.com/developer/apps/generate) in the store but I don't see any UI to modify that, and the URLs within info.xml seem not to be used by the store although they are used within the instance (/settings/apps). Any idea how I may add URLs and images?

**Comment by nursoda** [comment]
2.0.1 is released and its max-version is set to 22.

**Comment by andyxheli** [comment]
@nursoda Thank you so much! Looking forwarded to your improvements :)

**Comment by nursoda** [comment]
@andyxheli To be honest, I'm not a dev, just a nc admin. I merely try to care for that app as I thought it needed more timely updates. @all: Please propose code changes, I'll happily review them :))

**Comment by ChristophWurst** [comment]
> I have two more questions

1) Caching
2) https://apps.nextcloud.com/schema/apps/info.xsd defines what the info.xml can contain. There are elements for description, bug tracker and so on.

**Comment by nursoda** [comment]
@ChristophWurst Thanks for all your help!

1. Yes, worked over time, mine are all current now.
2. info.xml/store:
 
There are some points that I don't understand yet since my issue URL `<bugs>https://github.com/nursoda/twofactor_email/issues</bugs>` in info.xml is shown within the NC instance's `/settings/apps` but not in the store. But I'll deduct it from [Cospend](https://apps.nextcloud.com/apps/cospend) (where the issue URL is shown). And other parts I'll deduct from your [U2F info.xml](https://github.com/nextcloud/twofactor_u2f/blob/master/appinfo/info.xml), e.g. the screenshots.

What I don't understand: Why do you have these lines in your [U2F info.xml](https://github.com/nextcloud/twofactor_u2f/blob/master/appinfo/info.xml) and what impact does it have to add respectively to not have those lines, like: Are there cases where NC is set up without database? Should an app care about the DB type? Why does the app work fine with and without the activity definitions? Why is it safer to (not) provide the activities explicitly?

```
1.
        <dependencies>
                <database>sqlite</database>
                <database>mysql</database>
                <database>pgsql</database>
        </dependencies>
2.
        <activity>
                <settings>
                        <setting>OCA\TwoFactorU2F\Activity\Setting</setting>
                </settings>
                <providers>
                        <provider>OCA\TwoFactorU2F\Activity\Provider</provider>
                </providers>
        </activity>
```

I furthermore see a [dependabot security alert](https://github.com/nursoda/twofactor_email/security/dependabot/):

```
Dependabot cannot update *set-value* to a non-vulnerable version
The latest possible version that can be installed is 2.0.1 because of the following conflicting dependencies:
@nextcloud/axios@1.6.0 requires set-value@^2.0.0 via a transitive dependency on cache-base@1.0.1
The earliest fixed version is 4.0.1.
```

This results from a package you maintain: `nextcloud/axios`. It uses `babel/cli` as (full, not dev-) dependency – and potentially executes vulnerable code:

```
$ npm list set-value
twofactor_email@2.1.0
└─┬ @nextcloud/axios@1.6.0
  └─┬ @babel/cli@7.15.4
    └─┬ @nicolo-ribaudo/chokidar-2@2.1.8-no-fsevents.2
      └─┬ braces@2.3.2
        └─┬ snapdragon@0.8.2
          └─┬ base@0.11.2
            └─┬ cache-base@1.0.1
              ├── set-value@2.0.1
```

Can you tell whether this has to be a full dependency? In my package, babel is a devDependency only. If not, please move the dependency to a devDependency. Shall I create a PR for that or will you just fix it? If it IS a (full) dependency: Can you tell whether that (babel) *code* is actually used by axios in runtime?

BTW, I already contacted the babel maintainer nicolo-ribaudo, and he pointed out, that [explained](https://github.com/nicolo-ribaudo/chokidar-2/issues/8) why his module is that way, and that the code is not used. That is why I justified (`Dismiss`) the security alert as `Vulnerable code is not actually used`.

**Comment by ChristophWurst** [comment]
Please put your questions to https://help.nextcloud.com/c/dev/app-dev/24 as separate topics so we can hopefully help you collectively.
